x/evm/state/state.go

```diff
@@ -0,0 +1,127 @@
+package state
+
+import (
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/sei-protocol/sei-chain/x/evm/types"
+)
+
+func (s *StateDBImpl) CreateAccount(acc common.Address) {
+	// clear any existing state but keep balance untouched
+	s.clearAccountState(acc)
+
+	s.created[acc.String()] = struct{}{}
+	delete(s.selfDestructedAccs, acc.String())
+}
+
+func (s *StateDBImpl) GetCommittedState(addr common.Address, hash common.Hash) common.Hash {
+	return s.getState(addr, hash, func(kv sdk.KVStore, h common.Hash) []byte { return kv.GetCommitted(h[:]) })
+}
+
+func (s *StateDBImpl) GetState(addr common.Address, hash common.Hash) common.Hash {
+	return s.getState(addr, hash, func(kv sdk.KVStore, h common.Hash) []byte { return kv.Get(h[:]) })
+}
+
+func (s *StateDBImpl) getState(addr common.Address, hash common.Hash, getter func(sdk.KVStore, common.Hash) []byte) common.Hash {
+	val := getter(s.prefixStore(addr), hash)
+	if val == nil {
+		return common.Hash{}
+	}
+	return common.BytesToHash(val)
+}
+
+func (s *StateDBImpl) SetState(addr common.Address, key common.Hash, val common.Hash) {
+	s.prefixStore(addr).Set(key[:], val[:])
+}
+
+func (s *StateDBImpl) GetTransientState(addr common.Address, key common.Hash) common.Hash {
+	if addrState, ok := s.transientStorage[addr.String()]; !ok {
+		return common.Hash{}
+	} else if val, ok := addrState[key.String()]; !ok {
+		return common.Hash{}
+	} else {
+		return common.BytesToHash(val)
+	}
+}
+
+func (s *StateDBImpl) SetTransientState(addr common.Address, key, value common.Hash) {
+	addrKey := addr.String()
+	if addrState, ok := s.transientStorage[addrKey]; !ok {
+		s.transientStorage[addrKey] = map[string][]byte{key.String(): value[:]}
+	} else {
+		addrState[key.String()] = value[:]
```

Comment:
Might be good to check if prevValue is the same as current value and if so, just return: https://github.com/ethereum/go-ethereum/blame/614804b33c340cd60cbb6087d898b2968b0da320/core/state/statedb.go#L475

Also, do you need to make a copy of `value` here?

Reply:
updated it to use the same KV store as other states, and purge it at the end of the transaction https://github.com/sei-protocol/sei-chain/pull/1071/files#diff-ba4a9dd61688ef34761fc3792f6ef4a8bbf5f2a58d235e6d998928a154f9ef92R38